uasc/secure_channel.go

```diff
@@ -114,258 +94,202 @@ func NewSecureChannel(endpoint string, c *uacp.Conn, cfg *Config) (*SecureChanne
 	}
 
 	// Force the security mode to None if the policy is also None
+	// TODO: I don't like that a SecureChannel changes the incoming config
 	if cfg.SecurityPolicyURI == ua.SecurityPolicyURINone {
 		cfg.SecurityMode = ua.MessageSecurityModeNone
 	}
 
-	return &SecureChannel{
-		EndpointURL: endpoint,
+	s := &SecureChannel{
+		endpointURL: endpoint,
 		c:           c,
 		cfg:         cfg,
-		reqhdr: &ua.RequestHeader{
-			TimeoutHint:      uint32(cfg.RequestTimeout / time.Millisecond),
-			AdditionalHeader: ua.NewExtensionObject(nil),
-		},
-		state:     secureChannelCreated,
-		handler:   make(map[uint32]chan Response),
-		chunks:    make(map[uint32][]*MessageChunk),
-		requestID: cfg.RequestIDSeed,
-	}, nil
-}
+		requestID:   cfg.RequestIDSeed,
+	}
 
-func (s *SecureChannel) LocalEndpoint() string {
-	return s.EndpointURL
-}
+	s.reset()
 
-func (s *SecureChannel) Lifetime() uint32 {
-	return s.lifetime
+	return s, nil
 }
 
-func (s *SecureChannel) setState(n int32) {
-	atomic.StoreInt32(&s.state, n)
+func (s *SecureChannel) reset() {
```

Comment:
Why do we re-use a secure channel instance? I'm sure there is a reason but this deviates from the way normal Go programs function so it would be good to explain why.

Reply:
I think part of the issue is that we expose `secchan.Open` as an instance function rather than a static. This means the API allows use to call `s.Open()`, `s.Close()` and then `s.Open()` again.

Not the best API. Maybe worth removing/changing.